pandas/tests/groupby/test_groupby.py

```diff
@@ -2116,6 +2116,38 @@ def interweave(list_obj):
             exp = DataFrame({'key': keys, 'val': _exp_vals})
             assert_frame_equal(result, exp)
 
+    @pytest.mark.parametrize("test_series", [True, False])
+    @pytest.mark.parametrize("periods,fill_method,limit", [
+        (1, 'ffill', None), (1, 'ffill', 1),
+        (1, 'bfill', None), (1, 'bfill', 1),
+        (-1, 'ffill', None), (-1, 'ffill', 1),
+        (-1, 'bfill', None), (-1, 'bfill', 1)])
+    def test_pct_change(self, test_series, periods, fill_method, limit):
```

Comment:
It's worth calling out that the expectations I set up here are different than what was previously being returned by the GroupBy operation. Before this would return a MultiIndexed DataFrame, with the `key` column as the first level and the original index as the next level(s).

I couldn't find any existing test case or issue explicitly calling for that MultiIndex, so I took some liberty to consolidate the implementation and make the return value mimic what `shift` would return.

Reply:
Can you show an example of how a `MultiIndex` was being created?  Our 'rules' for return shapes on groupby transforms / apply  are kind of dark corner - as you're working through these might be worth thinking about formalizing them. xref to #5755, #6944 (last bullet), #5764, probably some others.

Reply:
Given
```python
df = pd.DataFrame({'key': ['a'] * 3 + ['b'] * 3, 'val': range(6)})
df.groupby('key').pct_change()
```

Previous behavior returned:
```python
            val
key            
a   0       NaN
    1       inf
    2  1.000000
b   3       NaN
    4  0.333333
    5  0.250000
```

After this change it returns:
```python
Out[3]: 
        val
0       NaN
1       inf
2  1.000000
3  0.500000
4  0.333333
5  0.250000
```

Which matches `shift` behavior before/after. Thanks for the linked notes - will give them a look

Reply:
hmm, something has changed in master, on `0.22` I get your new behavior
```python
df.groupby('key').pct_change()
        val
0       NaN
1       inf
2  1.000000
3       NaN
4  0.333333
5  0.250000

In [62]: pd.__version__
Out[62]: '0.22.0'
```

Reply:
so the return values (e.g. whether it includes the group keys or not), IOW is it a transforming operation. So shift/fill is like ``.transform('mean')``. ``pct_change`` should be the same I agree. Prob need some doc-updates http://pandas-docs.github.io/pandas-docs-travis/groupby.html#aggregation (and maybe next section), as well as an example in the whatsnew.

So ``.apply`` by default DOES return the multi-index and that is why you were getting this previously.

Reply:
`fill` is still different in this case. To sum up the existing behavior of those on master on a GroupBy:

 - `shift` does not include group keys in result
 - `pct_change` inserts the group keys into the first level(s) of the index AND
- `fill` includes the group keys as a column in the data of the result

@jreback two questions:
 1. If `shift` and `pct_change` return a like result, do we want `fill` to follow the same pattern?
 2. Do you have an example of where `.apply` is returning a multi-index? Doing apply with an anonymous func returns the group keys in a column not in the index:

```python
df.groupby('key').apply(lambda x: x * 5)
     key  val
0  aaaaa    0
1  aaaaa    5
2  aaaaa   10
3  bbbbb   15
4  bbbbb   20
5  bbbbb   25
```